### PR TITLE
Initial CR

### PR DESCRIPTION
**Pull request by bakks**

### CODE REVIEW

File: ext_shared_hashmap.cpp [hunk 1]
```diff
@@ -1,135 +0,0 @@
-/*
```
Comment (bakks) [inline ext_shared_hashmap.cpp]: Can drop this comment block

File: ext_shared_hashmap.cpp [hunk 2]
```diff
@@ -1,135 +0,0 @@
...
-      return Variant(String(value));
-    } catch (const std::out_of_range& oor) {
-      return Variant(Variant::NullInit{});
-    }
-  }
-
-  bool erase(const String& key) {
-    WriteLock write_lock(hashmap_mutex);
-
-    return hashmap.erase(key.toCppString());
```
Comment (bakks) [inline ext_shared_hashmap.cpp]: Does this throw an exception if the key doesn't exist?

File: ext_shared_hashmap.cpp [hunk 3]
```diff
@@ -1,135 +0,0 @@
...
-  }
-};
-
-static ReadWriteMutex shared_hashmap_mutex;
-static std::unordered_map<std::string, SharedHashMap*> shared_hashmaps;
-
-static bool HHVM_FUNCTION(shhashmap_init, const String& name) {
-  const std::string cpp_name = name.toCppString();
-
-  {
```
Comment (bakks) [inline ext_shared_hashmap.cpp]: Why do you need this first block?

File: ext_shared_hashmap.cpp [hunk 4]
```diff
@@ -1,135 +0,0 @@
...
-
-  bool erase(const String& key) {
-    WriteLock write_lock(hashmap_mutex);
-
-    return hashmap.erase(key.toCppString());
-  }
-};
-
-static ReadWriteMutex shared_hashmap_mutex;
-static std::unordered_map<std::string, SharedHashMap*> shared_hashmaps;
```
Comment (bakks) [inline ext_shared_hashmap.cpp]: Could you drop this intermediate hashmap if you used a single `SharedHashedMap` and prefixed the keys? You might have more lock contention in that case but we probably won't have a lot of hash maps.

**Comment by bakks** [comment]
cc @RyanGordon

File: ext_shared_hashmap.cpp (hunk 2 above)
Comment (RyanGordon) [inline ext_shared_hashmap.cpp]: No http://www.cplusplus.com/reference/map/map/erase/

File: ext_shared_hashmap.cpp (hunk 4 above)
Comment (RyanGordon) [inline ext_shared_hashmap.cpp]: No, you can't do that; you could have key collisions if you did that

File: ext_shared_hashmap.cpp (hunk 3 above)
Comment (RyanGordon) [inline ext_shared_hashmap.cpp]: This is a C construct. Variables are local to the block and at the end classes are destructed e.g. the read_lock.

Comment (bakks) [inline ext_shared_hashmap.cpp]: I mean why isn't it sufficient just to have the second block?

Comment (RyanGordon) [inline ext_shared_hashmap.cpp]: Because then you'll have an infinite wait

**Pull request closed without merging**